source/common/config/xds_mux/grpc_mux_impl.cc

```diff
@@ -0,0 +1,403 @@
+  return absl::nullopt;
+}
+
+template class GrpcMuxImpl<DeltaSubscriptionState, DeltaSubscriptionStateFactory,
+                           envoy::service::discovery::v3::DeltaDiscoveryRequest,
+                           envoy::service::discovery::v3::DeltaDiscoveryResponse>;
+template class GrpcMuxImpl<SotwSubscriptionState, SotwSubscriptionStateFactory,
+                           envoy::service::discovery::v3::DiscoveryRequest,
+                           envoy::service::discovery::v3::DiscoveryResponse>;
+
+// Delta- and SotW-specific concrete subclasses:
+GrpcMuxDelta::GrpcMuxDelta(Grpc::RawAsyncClientPtr&& async_client, Event::Dispatcher& dispatcher,
+                           const Protobuf::MethodDescriptor& service_method,
+                           envoy::config::core::v3::ApiVersion transport_api_version,
+                           Random::RandomGenerator& random, Stats::Scope& scope,
+                           const RateLimitSettings& rate_limit_settings,
+                           const LocalInfo::LocalInfo& local_info, bool skip_subsequent_node)
+    : GrpcMuxImpl(std::make_unique<DeltaSubscriptionStateFactory>(dispatcher), skip_subsequent_node,
+                  local_info, transport_api_version, std::move(async_client), dispatcher,
+                  service_method, random, scope, rate_limit_settings) {}
+
+// GrpcStreamCallbacks for GrpcMuxDelta
+void GrpcMuxDelta::requestOnDemandUpdate(const std::string& type_url,
+                                         const absl::flat_hash_set<std::string>& for_update) {
+  auto& sub = subscriptionStateFor(type_url);
+  sub.updateSubscriptionInterest(for_update, {});
+  // Tell the server about our change in interest, if any.
+  if (sub.subscriptionUpdatePending()) {
+    trySendDiscoveryRequests();
+  }
+}
+
+GrpcMuxSotw::GrpcMuxSotw(Grpc::RawAsyncClientPtr&& async_client, Event::Dispatcher& dispatcher,
+                         const Protobuf::MethodDescriptor& service_method,
+                         envoy::config::core::v3::ApiVersion transport_api_version,
+                         Random::RandomGenerator& random, Stats::Scope& scope,
+                         const RateLimitSettings& rate_limit_settings,
+                         const LocalInfo::LocalInfo& local_info, bool skip_subsequent_node)
+    : GrpcMuxImpl(std::make_unique<SotwSubscriptionStateFactory>(dispatcher), skip_subsequent_node,
+                  local_info, transport_api_version, std::move(async_client), dispatcher,
+                  service_method, random, scope, rate_limit_settings) {}
+
+Config::GrpcMuxWatchPtr NullGrpcMuxImpl::addWatch(const std::string&,
+                                                  const absl::flat_hash_set<std::string>&,
+                                                  SubscriptionCallbacks&, OpaqueResourceDecoder&,
+                                                  const SubscriptionOptions&) {
+  throw EnvoyException("ADS must be configured to support an ADS config source");
```

Comment:
nit: move this into the header file (the implementation of the entire NullGrpcMuxImpl will be there).

Reply:
I don't think we can throw in headers though :/

Reply:
This implementation is due to the "no-throwing-exceptions-in-headers" rule, which is enforced by our CI.